### PR TITLE
Show progress bar when exporting emails

### LINKED ISSUE #5546: Show progress when exporting emails
As a user who exports emails i want to see progress of the export action so i can estimate how long the action takes

**AC**
* Progress is shown in a blocking modal dialg.
* Dialog shows number of emails to export
* Dialog shows number of already exported email
* Dialog shows progress bar
* There is an option to cancel the export action

## Notes

Reported by a customer.

We don't show any progress when exporting emails so it is hard to tell what exactly is going on but the network requests have stoppped and the job didn't finish in a few hours.

We should at least show the actual progress, possibly with both text and progress bar.

## Test Notes
* [ ] Exporting a batch of emails show a dialog with a progress bar, the total of processed emails and the total emails count
* [ ] Exporting one single email still working without the progress bar
* [ ] Select and drag a batch of emails still working on desktop clients
* [ ] If one exporting email fails, the dialog is dismissed and no file is downloaded

### PR DESCRIPTION
**Pull request by murilopereirame**
When exporting mails no progress is informed to the user, just a blocking dialog asking to wait.

Now it show a progress bar and informs the user how many emails have been prepared and how many emails are going to be exported.

Co-Authors: @ganthern @rezbyte 

close #5546

### CODE REVIEW

**Comment by charlag** [comment]
While it is a great improvement it doesn't  cover everything in #5546. Ideally we should have a cancel button and export what we already downloaded. Should we split the issue?

**Comment by ganthern** [comment]
I think splitting is a good idea, but cancellation should be in this one probably (don't know how easy that is atm with the progress dialog)

File: src/mail/export/Exporter.ts [hunk 1]
```diff
@@ -65,16 +65,36 @@ export async function exportMails(
...
 	fileController: FileController,
 	operationId?: OperationId,
+	signal?: AbortSignal,
 ): Promise<void> {
 	// Considering that the effort for generating the bundle is higher
 	// than generating the files, we need to consider it twice, so the
 	// total effort would be (mailsToBundle * 2) + filesToGenerate
 	const totalMails = mails.length * 3
 	let doneMails = 0
+	let cancel = false
```
Comment (charlag) [inline src/mail/export/Exporter.ts:75]: maybe `cancelled`?

File: src/mail/export/Exporter.ts [hunk 2]
```diff
@@ -65,16 +65,36 @@ export async function exportMails(
...
+
+	const onAbort = () => {
+		cancel = true
+	}
+
+	//We use once here to let the AbortController remove the
+	//listener after dispatch the abort function
+	signal?.addEventListener("abort", onAbort, { once: true })
```
Comment (charlag) [inline src/mail/export/Exporter.ts:83]: As a general note: please don't throw something that doesn't extend `Error`.
We also have have `CancelledError` already.

I would reorganize a bit where this error is caught though. I would catch it somewhere inside this function and return  early instead. Otherwise it's a bit far from where it is thrown and it's hard to follow.

With that you can also do `removeEventListener` inside `finally` block

File: src/gui/dialogs/ProgressDialog.ts [hunk 3]
```diff
@@ -26,24 +29,27 @@ export async function showProgressDialog<T>(
 
 	const progressDialog = new Dialog(DialogType.Progress, {
 		view: () =>
-			m(
-				".hide-outline",
-				{
-					// We make this element focusable so that the screen reader announces the dialog
-					tabindex: TabIndex.Default,
+			m("", [
```
[the remark this answers is not in the document]
Comment (murilopereirame) [reply src/gui/dialogs/ProgressDialog.ts:32]: I think both options are good and stay with the pattern that we adopted to the app. In login during the 2FA auth we have both styles, the waiting dialog with the header and the dialog with a button

[the remark this answers is not in the document]
Comment (charlag) [reply src/gui/dialogs/ProgressDialog.ts:32]: ah yes, good point, let's keep it this way then

File: src/mail/export/Exporter.ts [hunk 4]
```diff
@@ -58,19 +60,66 @@ export function generateExportFileName(subject: string, sentOn: Date, mode: Mail
...
+			const mailFile = await generateMailFile(bundle, generateExportFileName(bundle.subject, new Date(bundle.receivedOn), mode), mode)
+			dataFiles.push(mailFile)
+			updateProgress()
+		}
+
+		const zipName = `${sortableTimestamp()}-${mode}-mail-export.zip`
+		const outputFile = await (dataFiles.length === 1 ? dataFiles[0] : zipDataFiles(dataFiles, zipName))
+		return fileController.saveDataFile(outputFile)
+	} catch (e) {
+		if (e.name !== "CancelledError") throw e
```
Comment (charlag) [inline src/mail/export/Exporter.ts:119]: I  think we normally do `instanceof` checks but this is also fine

**Review by charlag — APPROVED** [review]
